modules/core/src/main/java/com/google/refine/importers/ImporterUtilities.java

```diff
@@ -198,6 +198,26 @@ static public void setupColumns(Project project, List<String> columnNames) {
         project.columnModel.update();
     }
 
+    /**
+     * @param columnsHasData
+     *            Record if there is data in each column( false:null;true:has data)
+     */
+    static public void deleteEmptyColumns(Project project, List<Boolean> columnsHasData) throws ModelException {
+        // skip, if no columns to delete
+        if (columnsHasData.isEmpty() || columnsHasData.stream().allMatch(Boolean.TRUE::equals)) {
+            return;
+        }
+
```

Comment:
```suggestion
```

This extra pass is unnecessary, I think, since it'll be handled naturally by your loop below.

Reply:
* a complete extra pass is only in worst case, as `allMatch` will short circuit
* added it, because in case there is no column to delete, the `ColumnModel.update` method would still be called twice adding unnecessary calculations